strawberry/permission.py

```diff
@@ -1,21 +1,115 @@
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Any, Awaitable, Optional, Union
+import abc
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Awaitable,
+    Callable,
+    List,
+    Optional,
+    Type,
+    Union,
+    cast,
+)
+
+from graphql import GraphQLError
+
+from strawberry.exceptions import StrawberryGraphQLError
+from strawberry.extensions import FieldExtension
+from strawberry.schema_directive import Location, StrawberrySchemaDirective
+from strawberry.utils.await_maybe import await_maybe
 
 if TYPE_CHECKING:
-    from strawberry.types.info import Info
+    from strawberry.field import StrawberryField
+    from strawberry.types import Info
 
 
-class BasePermission:
+class BasePermission(abc.ABC):
     """
     Base class for creating permissions
     """
 
     message: Optional[str] = None
 
+    error_extensions: Optional[dict[str, str]] = None
+
+    error_class: Type[GraphQLError] = StrawberryGraphQLError
+
+    _schema_directive: Optional[StrawberrySchemaDirective] = None
+
+    @abc.abstractmethod
     def has_permission(
         self, source: Any, info: Info, **kwargs
     ) -> Union[bool, Awaitable[bool]]:
         raise NotImplementedError(
             "Permission classes should override has_permission method"
         )
+
+    def raise_error(self) -> None:
```

Comment:
An idea here: Instead of this being a function that should raise an error only, it could also be able to return something else?

Basically to solve [this issue](https://github.com/strawberry-graphql/strawberry/issues/2406), but also for some use cases I have

Reply:
What exactly would you like to return here? We would have to implement some mechanism to expose the type of that value from the permission to modify the `field`'s return type in `PermissionExtension::apply`. For typed errors, this is easy but supporting _any_ return is making things more complicated (e.g. parsing the type annotation of `Permission::has_permission`).

Reply:
Thanks for the clarification! While seeing the use-case for this, I'm also concerned that this will encourage the use of returning `None` for permissions. Being explicit about errors is a better approach in many cases and follows the `failure-first` approach. I believe it's better to focus on the core permissions feature here, because a user can easily extend the PermissionsExtension to accomplish this. To make that easier, I can move the error raising to the extension.

Typed errors should be handled in a separate Extension, as this extension is focused on classic `GQLError`-style permission errors. However, that would mean having to create another base class for permissions. An alternative is to introduce another layer: `BasePermission -> GQLErrorPermission alias Permission | TypedErrorPermission`. That way we could still handle both in the default extension. 

Please let me know what you think ☺️